protoc-gen-go/golden_test.go

```diff
@@ -123,3 +111,181 @@ func TestGolden(t *testing.T) {
 }
 
 var fdescRE = regexp.MustCompile(`(?ms)^var fileDescriptor.*}`)
+
+func TestParameters(t *testing.T) {
+	for _, test := range []struct {
+		parameters  string
+		wantFiles   map[string]bool
+		wantImports map[string]bool
+		goPackage   string
+	}{
+		{
+			parameters: "",
+			wantFiles: map[string]bool{
+				"package/alpha/a.pb.go": true,
+				"beta/b.pb.go":          true,
+			},
+			wantImports: map[string]bool{
+				"github.com/golang/protobuf/proto": true,
+				"beta": true,
+			},
+		},
+		{
+			parameters: "import_prefix=prefix",
+			wantFiles: map[string]bool{
+				"package/alpha/a.pb.go": true,
+				"beta/b.pb.go":          true,
+			},
+			wantImports: map[string]bool{
+				// This really doesn't seem like useful behavior.
+				"prefixgithub.com/golang/protobuf/proto": true,
+				"prefixbeta":                             true,
+			},
+		},
+		{
+			// import_path only affects the 'package' line.
+			parameters: "import_path=import/path/of/pkg",
+			wantFiles: map[string]bool{
+				"package/alpha/a.pb.go": true,
+				"beta/b.pb.go":          true,
+			},
+		},
+		{
+			parameters: "Mbeta/b.proto=package/gamma",
+			wantFiles: map[string]bool{
+				"package/alpha/a.pb.go": true,
+				"beta/b.pb.go":          true,
+			},
+			wantImports: map[string]bool{
+				"github.com/golang/protobuf/proto": true,
+				// Rewritten by the M parameter.
+				"package/gamma": true,
+			},
+		},
+		{
+			parameters: "import_prefix=prefix,Mbeta/b.proto=package/gamma",
+			wantFiles: map[string]bool{
+				"package/alpha/a.pb.go": true,
+				"beta/b.pb.go":          true,
+			},
+			wantImports: map[string]bool{
+				// import_prefix applies after M.
+				"prefixpackage/gamma": true,
+			},
+		},
+	} {
+		name := test.parameters
+		if name == "" {
+			name = "defaults"
+		}
+		// TODO: Switch to t.Run when we no longer support Go 1.6.
+		t.Logf("TEST: %v", name)
+		workdir, err := ioutil.TempDir("", "proto-test")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer os.RemoveAll(workdir)
+
+		for _, dir := range []string{"alpha", "beta", "out"} {
+			if err := os.MkdirAll(filepath.Join(workdir, dir), 0777); err != nil {
+				t.Fatal(err)
+			}
+		}
+
+		aProto := []byte(`
```

Comment:
Move these into constants at top of function? It's easier to understand the test cases after seeing these two proto files.

Reply:
done